[Bug]: Unable to compile DTLS1.3 SCTP support without TLS1.2

### Contact Details

nep-gh@packageloss.eu

### Version

v5.6.4-stable

### Description

commandline:
./configure --enable-dtls --enable-tls13 --enable-dtls13 --enable-sctp --disable-tlsv12

gcc version 13.2.0 (Ubuntu 13.2.0-4ubuntu3) 

/etc/os-release 
PRETTY_NAME="Ubuntu 23.10"
NAME="Ubuntu"
VERSION_ID="23.10"
VERSION="23.10 (Mantic Minotaur)"
VERSION_CODENAME=mantic
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=mantic
LOGO=ubuntu-logo

### Reproduction steps

Configure as above, then plainly compiling with make

Removing the  --disable-tlsv12 gives a working build.

### Relevant log output

```shell
make
make -j13  all-recursive
make[1]: Entering directory '/home/nephele/proj/crazy-network-lib/wolfssl'
make[2]: Entering directory '/home/nephele/proj/crazy-network-lib/wolfssl'
make[2]: warning: -j13 forced in submake: resetting jobserver mode.
  CC       examples/sctp/sctp-server.o
  CC       examples/sctp/sctp-client.o
  CC       examples/sctp/sctp-server-dtls.o
  CC       examples/sctp/sctp-client-dtls.o
  CC       examples/pem/pem.o
  CC       examples/asn1/asn1.o
  CC       wolfcrypt/test/testsuite_testsuite_test-test.o
  CC       examples/echoclient/testsuite_testsuite_test-echoclient.o
  CC       examples/client/testsuite_testsuite_test-client.o
  CC       examples/echoserver/testsuite_testsuite_test-echoserver.o
  CC       examples/server/testsuite_testsuite_test-server.o
  CC       testsuite/testsuite_test-testsuite.o
  CC       tests/unit_test-unit.o
examples/sctp/sctp-client-dtls.c: In function ‘main’:
examples/sctp/sctp-server-dtls.c: In function ‘main’:
examples/sctp/sctp-client-dtls.c:75:40: error: implicit declaration of function ‘wolfDTLSv1_2_client_method’; did you mean ‘wolfDTLSv1_3_client_method’? [-Werror=implicit-function-declaration]
   75 |     WOLFSSL_CTX* ctx = wolfSSL_CTX_new(wolfDTLSv1_2_client_method());
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        wolfDTLSv1_3_client_method
examples/sctp/sctp-client-dtls.c:75:40: error: nested extern declaration of ‘wolfDTLSv1_2_client_method’ [-Werror=nested-externs]
examples/sctp/sctp-server-dtls.c:81:40: error: implicit declaration of function ‘wolfDTLSv1_2_server_method’; did you mean ‘wolfDTLSv1_3_server_method’? [-Werror=implicit-function-declaration]
   81 |     WOLFSSL_CTX* ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method());
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        wolfDTLSv1_3_server_method
examples/sctp/sctp-server-dtls.c:81:40: error: nested extern declaration of ‘wolfDTLSv1_2_server_method’ [-Werror=nested-externs]
examples/sctp/sctp-client-dtls.c:75:40: error: passing argument 1 of ‘wolfSSL_CTX_new’ makes pointer from integer without a cast [-Werror=int-conversion]
   75 |     WOLFSSL_CTX* ctx = wolfSSL_CTX_new(wolfDTLSv1_2_client_method());
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        |
      |                                        int
examples/sctp/sctp-server-dtls.c:81:40: error: passing argument 1 of ‘wolfSSL_CTX_new’ makes pointer from integer without a cast [-Werror=int-conversion]
   81 |     WOLFSSL_CTX* ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method());
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        |
      |                                        int
In file included from examples/sctp/sctp-server-dtls.c:27:
./wolfssl/ssl.h:1127:70: note: expected ‘WOLFSSL_METHOD *’ but argument is of type ‘int’
 1127 | WOLFSSL_ABI WOLFSSL_API WOLFSSL_CTX* wolfSSL_CTX_new(WOLFSSL_METHOD* method);
      |                                                      ~~~~~~~~~~~~~~~~^~~~~~
In file included from examples/sctp/sctp-client-dtls.c:28:
./wolfssl/ssl.h:1127:70: note: expected ‘WOLFSSL_METHOD *’ but argument is of type ‘int’
 1127 | WOLFSSL_ABI WOLFSSL_API WOLFSSL_CTX* wolfSSL_CTX_new(WOLFSSL_METHOD* method);
      |                                                      ~~~~~~~~~~~~~~~~^~~~~~
  CC       tests/unit_test-api.o
cc1: all warnings being treated as errors
make[2]: *** [Makefile:6310: examples/sctp/sctp-server-dtls.o] Error 1
make[2]: *** Waiting for unfinished jobs....
cc1: all warnings being treated as errors
make[2]: *** [Makefile:6310: examples/sctp/sctp-client-dtls.o] Error 1
In file included from tests/api.c:152:
tests/api.c: In function ‘test_wolfSSL_dtls_set_mtu’:
tests/api.c:59428:41: error: implicit declaration of function ‘wolfDTLSv1_2_server_method’; did you mean ‘wolfDTLSv1_3_server_method’? [-Werror=implicit-function-declaration]
59428 |     ExpectNotNull(ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:59428:5: note: in expansion of macro ‘ExpectNotNull’
59428 |     ExpectNotNull(ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |     ^~~~~~~~~~~~~
tests/api.c:59428:41: error: nested extern declaration of ‘wolfDTLSv1_2_server_method’ [-Werror=nested-externs]
59428 |     ExpectNotNull(ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:59428:5: note: in expansion of macro ‘ExpectNotNull’
59428 |     ExpectNotNull(ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |     ^~~~~~~~~~~~~
tests/api.c:59428:41: error: passing argument 1 of ‘wolfSSL_CTX_new’ makes pointer from integer without a cast [-Werror=int-conversion]
59428 |     ExpectNotNull(ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                         |
      |                                         int
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:59428:5: note: in expansion of macro ‘ExpectNotNull’
59428 |     ExpectNotNull(ctx = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |     ^~~~~~~~~~~~~
In file included from tests/api.c:150:
./wolfssl/ssl.h:1127:70: note: expected ‘WOLFSSL_METHOD *’ but argument is of type ‘int’
 1127 | WOLFSSL_ABI WOLFSSL_API WOLFSSL_CTX* wolfSSL_CTX_new(WOLFSSL_METHOD* method);
      |                                                      ~~~~~~~~~~~~~~~~^~~~~~
tests/api.c: In function ‘test_wolfSSL_dtls_plaintext’:
tests/api.c:59548:10: error: ‘wolfDTLSv1_2_client_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_client_method’?
59548 |         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |          wolfDTLSv1_3_client_method
tests/api.c:59548:10: note: each undeclared identifier is reported only once for each function it appears in
tests/api.c:59548:38: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
59548 |         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                      wolfDTLSv1_3_server_method
tests/api.c: In function ‘test_wolfSSL_dtls_fragments’:
tests/api.c:59701:10: error: ‘wolfDTLSv1_2_client_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_client_method’?
59701 |         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |          wolfDTLSv1_3_client_method
tests/api.c:59701:38: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
59701 |         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                      wolfDTLSv1_3_server_method
tests/api.c: In function ‘_test_wolfSSL_ignore_alert_before_cookie’:
tests/api.c:59767:29: error: ‘wolfDTLSv1_2_client_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_client_method’?
59767 |         client_cbs.method = wolfDTLSv1_2_client_method;
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                             wolfDTLSv1_3_client_method
tests/api.c:59768:29: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
59768 |         server_cbs.method = wolfDTLSv1_2_server_method;
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                             wolfDTLSv1_3_server_method
tests/api.c: In function ‘test_wolfSSL_dtls_bad_record’:
tests/api.c:59864:41: error: ‘wolfDTLSv1_2_client_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_client_method’?
59864 |     ret = _test_wolfSSL_dtls_bad_record(wolfDTLSv1_2_client_method,
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                         wolfDTLSv1_3_client_method
tests/api.c:59865:9: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
59865 |         wolfDTLSv1_2_server_method);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |         wolfDTLSv1_3_server_method
tests/api.c: In function ‘test_wolfSSL_dtls_stateless’:
tests/api.c:60358:10: error: ‘wolfDTLSv1_2_client_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_client_method’?
60358 |         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |          wolfDTLSv1_3_client_method
tests/api.c:60358:38: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
60358 |         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                      wolfDTLSv1_3_server_method
tests/api.c: In function ‘test_dtls_ipv6_check’:
tests/api.c:65616:43: error: implicit declaration of function ‘wolfDTLSv1_2_client_method’; did you mean ‘wolfDTLSv1_3_client_method’? [-Werror=implicit-function-declaration]
65616 |     ExpectNotNull(ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method()));
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:65616:5: note: in expansion of macro ‘ExpectNotNull’
65616 |     ExpectNotNull(ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method()));
      |     ^~~~~~~~~~~~~
tests/api.c:65616:43: error: nested extern declaration of ‘wolfDTLSv1_2_client_method’ [-Werror=nested-externs]
65616 |     ExpectNotNull(ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method()));
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:65616:5: note: in expansion of macro ‘ExpectNotNull’
65616 |     ExpectNotNull(ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method()));
      |     ^~~~~~~~~~~~~
tests/api.c:65616:43: error: passing argument 1 of ‘wolfSSL_CTX_new’ makes pointer from integer without a cast [-Werror=int-conversion]
65616 |     ExpectNotNull(ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method()));
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                           |
      |                                           int
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:65616:5: note: in expansion of macro ‘ExpectNotNull’
65616 |     ExpectNotNull(ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method()));
      |     ^~~~~~~~~~~~~
./wolfssl/ssl.h:1127:70: note: expected ‘WOLFSSL_METHOD *’ but argument is of type ‘int’
 1127 | WOLFSSL_ABI WOLFSSL_API WOLFSSL_CTX* wolfSSL_CTX_new(WOLFSSL_METHOD* method);
      |                                                      ~~~~~~~~~~~~~~~~^~~~~~
tests/api.c:65618:43: error: passing argument 1 of ‘wolfSSL_CTX_new’ makes pointer from integer without a cast [-Werror=int-conversion]
65618 |     ExpectNotNull(ctx_s = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                           |
      |                                           int
./tests/unit.h:143:36: note: in definition of macro ‘Expect’
  143 |     if (_ret != TEST_FAIL) { if (!(test)) ExpFail(description, result);        \
      |                                    ^~~~
tests/api.c:65618:5: note: in expansion of macro ‘ExpectNotNull’
65618 |     ExpectNotNull(ctx_s = wolfSSL_CTX_new(wolfDTLSv1_2_server_method()));
      |     ^~~~~~~~~~~~~
./wolfssl/ssl.h:1127:70: note: expected ‘WOLFSSL_METHOD *’ but argument is of type ‘int’
 1127 | WOLFSSL_ABI WOLFSSL_API WOLFSSL_CTX* wolfSSL_CTX_new(WOLFSSL_METHOD* method);
      |                                                      ~~~~~~~~~~~~~~~~^~~~~~
tests/api.c: In function ‘test_dtls_client_hello_timeout_downgrade’:
tests/api.c:66278:37: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
66278 |             wolfDTLS_client_method, wolfDTLSv1_2_server_method), 0);
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:160:24: note: in definition of macro ‘ExpectInt’
  160 |         int _x = (int)(x);                                                     \
      |                        ^
tests/api.c:66277:9: note: in expansion of macro ‘ExpectIntEQ’
66277 |         ExpectIntEQ(test_memio_setup(&test_ctx, &ctx_c, &ctx_s, &ssl_c, &ssl_s,
      |         ^~~~~~~~~~~
tests/api.c: In function ‘test_dtls_dropped_ccs’:
tests/api.c:66485:9: error: ‘wolfDTLSv1_2_client_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_client_method’?
66485 |         wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method), 0);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:160:24: note: in definition of macro ‘ExpectInt’
  160 |         int _x = (int)(x);                                                     \
      |                        ^
tests/api.c:66484:5: note: in expansion of macro ‘ExpectIntEQ’
66484 |     ExpectIntEQ(test_memio_setup(&test_ctx, &ctx_c, &ctx_s, &ssl_c, &ssl_s,
      |     ^~~~~~~~~~~
tests/api.c:66485:37: error: ‘wolfDTLSv1_2_server_method’ undeclared (first use in this function); did you mean ‘wolfDTLSv1_3_server_method’?
66485 |         wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method), 0);
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
./tests/unit.h:160:24: note: in definition of macro ‘ExpectInt’
  160 |         int _x = (int)(x);                                                     \
      |                        ^
tests/api.c:66484:5: note: in expansion of macro ‘ExpectIntEQ’
66484 |     ExpectIntEQ(test_memio_setup(&test_ctx, &ctx_c, &ctx_s, &ssl_c, &ssl_s,
      |     ^~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [Makefile:7608: tests/unit_test-api.o] Error 1
make[2]: Leaving directory '/home/nephele/proj/crazy-network-lib/wolfssl'
make[1]: *** [Makefile:7979: all-recursive] Error 1
make[1]: Leaving directory '/home/nephele/proj/crazy-network-lib/wolfssl'
make: *** [Makefile:4707: all] Error 2
```

--- a/examples/sctp/sctp-client-dtls.c
+++ b/examples/sctp/sctp-client-dtls.c
@@ -27,7 +27,7 @@
 #include <wolfssl/wolfcrypt/settings.h>
 #include <wolfssl/ssl.h>
 
-#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS)
+#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12)
 /* sctp */
 #include <sys/socket.h>
 #include <sys/types.h>
@@ -47,13 +47,13 @@ static int err_sys(const char* msg)
     perror(msg);
     exit(EXIT_FAILURE);
 }
-#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS */
+#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS && !WOLFSSL_NO_TLS12 */
 
 int main(int argc, char **argv)
 {
     (void)argc;
     (void)argv;
-#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS)
+#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12)
     int sd = socket(PF_INET, SOCK_STREAM, IPPROTO_SCTP);
 
     if (sd < 0)
@@ -126,7 +126,7 @@ int main(int argc, char **argv)
     wolfSSL_CTX_free(ctx);
 
     close(sd);
-#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS */
+#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS && !WOLFSSL_NO_TLS12 */
 
     return 0;
 }

--- a/examples/sctp/sctp-server-dtls.c
+++ b/examples/sctp/sctp-server-dtls.c
@@ -26,7 +26,7 @@
 #include <wolfssl/wolfcrypt/settings.h>
 #include <wolfssl/ssl.h>
 
-#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS)
+#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12)
 /* sctp */
 #include <sys/socket.h>
 #include <sys/types.h>
@@ -47,13 +47,13 @@ static int err_sys(const char* msg)
     perror(msg);
     exit(EXIT_FAILURE);
 }
-#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS */
+#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS && !WOLFSSL_NO_TLS12 */
 
 int main(int argc, char **argv)
 {
     (void)argc;
     (void)argv;
-#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS)
+#if defined(WOLFSSL_SCTP) && defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12)
     int sd = socket(PF_INET, SOCK_STREAM, IPPROTO_SCTP);
 
     if (sd < 0)
@@ -125,6 +125,6 @@ int main(int argc, char **argv)
     wolfSSL_CTX_free(ctx);
 
     close(sd);
-#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS */
+#endif /* WOLFSSL_SCTP && WOLFSSL_DTLS && !WOLFSSL_NO_TLS12 */
     return 0;
 }

--- a/src/dtls.c
+++ b/src/dtls.c
@@ -868,6 +868,7 @@ static int SendStatelessReply(const WOLFSSL* ssl, WolfSSL_CH* ch, byte isTls13)
     else
 #endif
     {
+#if !defined(WOLFSSL_NO_TLS12)
         if (!ch->dtls12cookieSet) {
             ret = CreateDtls12Cookie(ssl, ch, ch->dtls12cookie);
             if (ret != 0)
@@ -876,6 +877,11 @@ static int SendStatelessReply(const WOLFSSL* ssl, WolfSSL_CH* ch, byte isTls13)
         }
         ret = SendHelloVerifyRequest((WOLFSSL*)ssl, ch->dtls12cookie,
                 DTLS_COOKIE_SZ);
+#else
+        WOLFSSL_MSG("DTLS1.2 disabled with WOLFSSL_NO_TLS12");
+        WOLFSSL_ERROR_VERBOSE(NOT_COMPILED_IN);
+        ret = NOT_COMPILED_IN;
+#endif
     }
     return ret;
 }

--- a/src/dtls13.c
+++ b/src/dtls13.c
@@ -372,8 +372,14 @@ int Dtls13ProcessBufferedMessages(WOLFSSL* ssl)
                 downgraded = 1;
         }
         else {
+#if !defined(WOLFSSL_NO_TLS12)
             ret = DoHandShakeMsgType(ssl, msg->fullMsg, &idx, msg->type,
                     msg->sz, msg->sz);
+#else
+            WOLFSSL_MSG("DTLS1.2 disabled with WOLFSSL_NO_TLS12");
+            WOLFSSL_ERROR_VERBOSE(NOT_COMPILED_IN);
+            ret = NOT_COMPILED_IN;
+#endif
         }
 
         /* processing certificate_request triggers a connect. The error came

--- a/src/tls.c
+++ b/src/tls.c
@@ -678,55 +678,6 @@ int wolfSSL_make_eap_keys(WOLFSSL* ssl, void* msk, unsigned int len,
     return ret;
 }
 
-int wolfSSL_GetHmacType_ex(CipherSpecs* specs)
-{
-    if (specs == NULL)
-        return BAD_FUNC_ARG;
-
-    switch (specs->mac_algorithm) {
-        #ifndef NO_MD5
-        case md5_mac:
-        {
-            return WC_MD5;
-        }
-        #endif
-        #ifndef NO_SHA256
-        case sha256_mac:
-        {
-            return WC_SHA256;
-        }
-        #endif
-        #ifdef WOLFSSL_SHA384
-        case sha384_mac:
-        {
-            return WC_SHA384;
-        }
-        #endif
-        #ifdef WOLFSSL_SM3
-        case sm3_mac:
-        {
-            return WC_SM3;
-        }
-        #endif
-        #ifndef NO_SHA
-        case sha_mac:
-        {
-            return WC_SHA;
-        }
-        #endif
-        #ifdef HAVE_BLAKE2
-        case blake2b_mac:
-        {
-            return BLAKE2B_ID;
-        }
-        #endif
-        default:
-        {
-            return WOLFSSL_FATAL_ERROR;
-        }
-    }
-}
-
 /* return HMAC digest type in wolfSSL format */
 int wolfSSL_GetHmacType(WOLFSSL* ssl)
 {
@@ -1276,6 +1227,55 @@ int TLS_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz, int padSz,
 #endif /* WOLFSSL_AEAD_ONLY */
 
 #endif /* !WOLFSSL_NO_TLS12 */
+
+int wolfSSL_GetHmacType_ex(CipherSpecs* specs)
+{
+    if (specs == NULL)
+        return BAD_FUNC_ARG;
+
+    switch (specs->mac_algorithm) {
+        #ifndef NO_MD5
+        case md5_mac:
+        {
+            return WC_MD5;
+        }
+        #endif
+        #ifndef NO_SHA256
+        case sha256_mac:
+        {
+            return WC_SHA256;
+        }
+        #endif
+        #ifdef WOLFSSL_SHA384
+        case sha384_mac:
+        {
+            return WC_SHA384;
+        }
+        #endif
+        #ifdef WOLFSSL_SM3
+        case sm3_mac:
+        {
+            return WC_SM3;
+        }
+        #endif
+        #ifndef NO_SHA
+        case sha_mac:
+        {
+            return WC_SHA;
+        }
+        #endif
+        #ifdef HAVE_BLAKE2
+        case blake2b_mac:
+        {
+            return BLAKE2B_ID;
+        }
+        #endif
+        default:
+        {
+            return WOLFSSL_FATAL_ERROR;
+        }
+    }
+}
 
 #ifdef HAVE_TLS_EXTENSIONS
 

--- a/tests/api.c
+++ b/tests/api.c
@@ -59684,7 +59684,8 @@ static int test_wolfSSL_dtls_set_mtu(void)
 {
     EXPECT_DECLS;
 #if (defined(WOLFSSL_DTLS_MTU) || defined(WOLFSSL_SCTP)) && \
-    !defined(NO_WOLFSSL_SERVER) && defined(WOLFSSL_DTLS)
+    !defined(NO_WOLFSSL_SERVER) && defined(WOLFSSL_DTLS) && \
+    !defined(WOLFSSL_NO_TLS12)
     WOLFSSL_CTX* ctx = NULL;
     WOLFSSL*     ssl = NULL;
     const char* testCertFile;
@@ -59722,7 +59723,7 @@ static int test_wolfSSL_dtls_set_mtu(void)
 }
 
 #if defined(HAVE_IO_TESTS_DEPENDENCIES) && !defined(SINGLE_THREADED) && \
-    defined(WOLFSSL_DTLS)
+    defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12)
 
 static WC_INLINE void generateDTLSMsg(byte* out, int outSz, word32 seq,
         enum HandShakeType hsType, word16 length)
@@ -59842,7 +59843,7 @@ static int test_wolfSSL_dtls_plaintext(void) {
 #endif
 
 #if defined(HAVE_IO_TESTS_DEPENDENCIES) && !defined(SINGLE_THREADED) && \
-    defined(WOLFSSL_DTLS)
+    defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12)
 
 static void test_wolfSSL_dtls12_fragments_spammer(WOLFSSL* ssl)
 {
@@ -59963,8 +59964,10 @@ static int test_wolfSSL_dtls_fragments(void)
         method_provider server_meth;
         ssl_callback spammer;
     } params[] = {
+#if !defined(WOLFSSL_NO_TLS12)
         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
                 test_wolfSSL_dtls12_fragments_spammer},
+#endif
 #ifdef WOLFSSL_DTLS13
         {wolfDTLSv1_3_client_method, wolfDTLSv1_3_server_method,
                 test_wolfSSL_dtls13_fragments_spammer},
@@ -60029,10 +60032,15 @@ static int _test_wolfSSL_ignore_alert_before_cookie(byte version12)
     XMEMSET(&server_cbs, 0, sizeof(server_cbs));
     client_cbs.doUdp = server_cbs.doUdp = 1;
     if (version12) {
+#if !defined(WOLFSSL_NO_TLS12)
         client_cbs.method = wolfDTLSv1_2_client_method;
         server_cbs.method = wolfDTLSv1_2_server_method;
+#else
+        return TEST_SKIPPED;
+#endif
     }
-    else {
+    else
+    {
 #ifdef WOLFSSL_DTLS13
         client_cbs.method = wolfDTLSv1_3_client_method;
         server_cbs.method = wolfDTLSv1_3_server_method;
@@ -60125,17 +60133,18 @@ static int _test_wolfSSL_dtls_bad_record(
 
 static int test_wolfSSL_dtls_bad_record(void)
 {
-    int ret;
+    int ret = TEST_SUCCESS;
+#if !defined(WOLFSSL_NO_TLS12)
     ret = _test_wolfSSL_dtls_bad_record(wolfDTLSv1_2_client_method,
         wolfDTLSv1_2_server_method);
+#endif
 #ifdef WOLFSSL_DTLS13
-    if (ret != TEST_SUCCESS)
-        return ret;
-    return _test_wolfSSL_dtls_bad_record(wolfDTLSv1_3_client_method,
+    if (ret == TEST_SUCCESS) {
+        ret = _test_wolfSSL_dtls_bad_record(wolfDTLSv1_3_client_method,
         wolfDTLSv1_3_server_method);
-#else
-    return ret;
+    }
 #endif /* WOLFSSL_DTLS13 */
+    return ret;
 
 }
 
@@ -60620,8 +60629,10 @@ static int test_wolfSSL_dtls_stateless(void)
         ssl_callback client_ssl_ready;
         ssl_callback server_ssl_ready;
     } test_params[] = {
+#if !defined(WOLFSSL_NO_TLS12)
         {wolfDTLSv1_2_client_method, wolfDTLSv1_2_server_method,
                 test_wolfSSL_dtls_send_ch, test_wolfSSL_dtls_compare_stateless},
+#endif
 #if defined(WOLFSSL_DTLS13) && defined(WOLFSSL_SEND_HRR_COOKIE)
         {wolfDTLSv1_3_client_method, wolfDTLSv1_3_server_method,
                 test_wolfSSL_dtls_send_ch, test_wolfSSL_dtls_enable_hrrcookie},
@@ -60629,6 +60640,10 @@ static int test_wolfSSL_dtls_stateless(void)
                 test_wolfSSL_dtls_send_ch_with_invalid_cookie, test_wolfSSL_dtls_enable_hrrcookie},
 #endif
     };
+
+    if (0 == sizeof(test_params)){
+        return TEST_SKIPPED;
+    }
 
     for (i = 0; i < sizeof(test_params)/sizeof(*test_params); i++) {
         XMEMSET(&client_cbs, 0, sizeof(client_cbs));
@@ -65867,7 +65882,7 @@ static int test_dtls_msg_from_other_peer(void)
         *  !defined(SINGLE_THREADED) && !defined(NO_RSA) */
 #if defined(WOLFSSL_DTLS) && !defined(WOLFSSL_IPV6) &&               \
     !defined(NO_WOLFSSL_CLIENT) && !defined(NO_WOLFSSL_SERVER) &&   \
-    defined(HAVE_IO_TESTS_DEPENDENCIES)
+    defined(HAVE_IO_TESTS_DEPENDENCIES) && !defined(WOLFSSL_NO_TLS12)
 static int test_dtls_ipv6_check(void)
 {
     EXPECT_DECLS;
@@ -66017,7 +66032,8 @@ static int test_wolfSSL_configure_args(void)
 static int test_dtls_no_extensions(void)
 {
     EXPECT_DECLS;
-#if defined(WOLFSSL_DTLS) && defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES)
+#if defined(WOLFSSL_DTLS) && defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && \
+    !defined(WOLFSSL_NO_TLS12)
     WOLFSSL *ssl_s = NULL;
     WOLFSSL_CTX *ctx_s = NULL;
     struct test_memio_ctx test_ctx;
@@ -66498,7 +66514,9 @@ static int test_dtls_downgrade_scr(void)
 }
 #endif
 
-#if defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && defined(WOLFSSL_DTLS13)
+#if defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && defined(WOLFSSL_DTLS13) \
+    && !defined(WOLFSSL_NO_TLS12)
+
 static int test_dtls_client_hello_timeout_downgrade_read_cb(WOLFSSL *ssl,
         char *data, int sz, void *ctx)
 {
@@ -66525,7 +66543,9 @@ static int test_dtls_client_hello_timeout_downgrade_read_cb(WOLFSSL *ssl,
 static int test_dtls_client_hello_timeout_downgrade(void)
 {
     EXPECT_DECLS;
-#if defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && defined(WOLFSSL_DTLS13)
+#if defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && defined(WOLFSSL_DTLS13) \
+    && !defined(WOLFSSL_NO_TLS12)
+
     WOLFSSL_CTX *ctx_c = NULL;
     WOLFSSL_CTX *ctx_s = NULL;
     WOLFSSL *ssl_c = NULL;
@@ -66733,7 +66753,9 @@ static int test_dtls_client_hello_timeout(void)
 static int test_dtls_dropped_ccs(void)
 {
     EXPECT_DECLS;
-#if defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && defined(WOLFSSL_DTLS)
+#if defined(HAVE_MANUAL_MEMIO_TESTS_DEPENDENCIES) && defined(WOLFSSL_DTLS) \
+    && !defined(WOLFSSL_NO_TLS12)
+
     WOLFSSL_CTX *ctx_c = NULL;
     WOLFSSL_CTX *ctx_s = NULL;
     WOLFSSL *ssl_c = NULL;

